pkg/graph/constructor.go

```diff
@@ -17,14 +17,154 @@ limitations under the License.
 package graph
 
 import (
+	"context"
 	"fmt"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
 	eventingv1 "knative.dev/eventing/pkg/apis/eventing/v1"
 	eventingv1beta3 "knative.dev/eventing/pkg/apis/eventing/v1beta3"
 	messagingv1 "knative.dev/eventing/pkg/apis/messaging/v1"
+	eventingclient "knative.dev/eventing/pkg/client/injection/client"
 	duckv1 "knative.dev/pkg/apis/duck/v1"
 )
 
+func ConstructGraph(ctx context.Context, filterFunc func(obj interface{}) bool) (*Graph, error) {
+	eventingClient := eventingclient.Get(ctx)
+
+	g := NewGraph()
+
+	brokers, err := eventingClient.EventingV1().Brokers("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, broker := range brokers.Items {
+		if filterFunc(broker) {
+			g.AddBroker(broker)
+		}
+	}
+
+	channels, err := eventingClient.MessagingV1().Channels("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, channel := range channels.Items {
+		if filterFunc(channel) {
+			g.AddChannel(channel)
+		}
+	}
+
+	apiServerSources, err := eventingClient.SourcesV1().ApiServerSources("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, apiServerSource := range apiServerSources.Items {
+		if filterFunc(apiServerSource) {
+			g.AddSource(duckv1.Source{
+				ObjectMeta: apiServerSource.ObjectMeta,
+				TypeMeta:   apiServerSource.TypeMeta,
+				Spec:       apiServerSource.Spec.SourceSpec,
+				Status:     apiServerSource.Status.SourceStatus,
+			})
+		}
+	}
+
+	containerSources, err := eventingClient.SourcesV1().ContainerSources("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, containerSource := range containerSources.Items {
+		if filterFunc(containerSource) {
+			g.AddSource(duckv1.Source{
+				ObjectMeta: containerSource.ObjectMeta,
+				TypeMeta:   containerSource.TypeMeta,
+				Spec:       containerSource.Spec.SourceSpec,
+				Status:     containerSource.Status.SourceStatus,
+			})
+		}
+	}
+
+	pingSources, err := eventingClient.SourcesV1().PingSources("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pingSource := range pingSources.Items {
+		if filterFunc(pingSource) {
+			g.AddSource(duckv1.Source{
+				ObjectMeta: pingSource.ObjectMeta,
+				TypeMeta:   pingSource.TypeMeta,
+				Spec:       pingSource.Spec.SourceSpec,
+				Status:     pingSource.Status.SourceStatus,
+			})
+		}
+	}
+
+	sinkBindings, err := eventingClient.SourcesV1().SinkBindings("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, sinkBinding := range sinkBindings.Items {
+		if filterFunc(sinkBinding) {
+			g.AddSource(duckv1.Source{
+				ObjectMeta: sinkBinding.ObjectMeta,
+				TypeMeta:   sinkBinding.TypeMeta,
+				Spec:       sinkBinding.Spec.SourceSpec,
+				Status:     sinkBinding.Status.SourceStatus,
+			})
+		}
+	}
```

Comment:
@pierDipi do you have any ideas how we could get the sources (& in the duckv1.Source struct) more easily? This process is rather manual, and would not detect any sources not in eventing core :(

Reply:
there is annotation/label on the CRDs that implement the source ducktype

https://github.com/knative/eventing/blob/c53b44d67c71260e34f769eaf09da72f52ef197d/config/core/resources/containersource.yaml#L20

from the CRD list (with label selector) using dynamic client you can get the actual resource list